Default cubemap renaming

dbampersand opened the pull request:
Fix for this issue: https://github.com/momentum-mod/lumper/issues/60

Didn't use ```CubemapRegex``` as matching cX_cY_cZ as well will cause false positives. (Unless I'm misunderstanding how that works)

tsa96 commented:
On phone so not doing proper review but this should probably live in Lumper.Lib code so CLI can use it. Actually sounds like quite a helpful use for the cli, renaming BSP without breaking cubemaps (I have a WIP branch atm with CLI support for saving new BSPs)

dbampersand commented:
Great - I've moved it to be in lib and fixed the UI state not changing.

tsa96 commented on src/Lumper.UI/Services/BspService.cs:
```diff
@@ -275,6 +278,12 @@ public async Task<bool> Save(IStorageFile? outFile)
             if (Program.Desktop.MainWindow is not null)
             {
                 var outName = outFile is not null ? Path.GetFileName(outFile.Path.AbsolutePath) : FileName;
+                if (outName != FileName)
+                {
+                    BspFile.GetLump<PakfileLump>().RenameDefaultCubemap(outName);
+                    BspService.Instance.PakfileLumpViewModel.IsModified = true;
+                    BspService.Instance.ResetLumpViewModels();
+                }
```
Still don't think this is the right approach, should do this logic in the BspFile SaveToFile method. Probably worth adding a checkbox to the File... menu to toggle this in UI though, then pass that into SaveToFile.

I'd like to avoid reloading lump viewmodels here, since that would cause every UI page to reload. Unfortunately I can't think of a way to update the viewmodels that isn't ugly, but still worth it to avoid reloading the whole things. After SaveToFile is called, if successful, iterate through every PakfileEntryViewModel and update their Key to the Key of the underlying PakfileEntry, if changed.

Don't love doing this but should be perfectly performant, and will work if there's ever other pakfile changes we make during saving in the future.

tsa96 commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.cs:
```diff
@@ -38,6 +39,30 @@ public class PakfileLump(BspFile parent) : ManagedLump<BspLumpType>(parent), IFi
 
     private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
 
+    //as per https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/utils/vbsp/cubemap.cpp
```
This doesn't seem right. Maps I'm looking at often have hundreds of cubemaps stored in `maps/<name>/c_<x coord>_<y coord>_<z coord>.vtf`, these should also be moved. grepping for "maps/" in that CPP file seems to find what I'm talking about: https://github.com/ValveSoftware/source-sdk-2013/blob/0d8dceea4310fde5706b3ce1c70609d72a38efdf/mp/src/utils/vbsp/cubemap.cpp#L512

dbampersand commented:
I've updated it for the logic to be in lib-only code and also fixed updating path references. Not confident on the way I've done updating the UI (the stuff in BspService), think I need to read more on how Avalonia actually works. I'll update later to clean up and I'll try to see if I can make the UI code less kludgy.

tsa96 commented on src/Lumper.Lib/BSP/BspFile.cs:
```diff
@@ -105,6 +105,15 @@ public sealed partial class BspFile : IDisposable
+            Dictionary<string, string> modified = GetLump<PakfileLump>().RenameDefaultCubemap(Path.GetFileName(path));
+            foreach (KeyValuePair<string,string> modifiedPath in modified)
+            {
+                GetLump<PakfileLump>().UpdatePathReferences(modifiedPath.Value, modifiedPath.Key);
```
Weird, UpdatePathReferences doesn't live in Lumper.Lib, I suspect you missed something from your commits.

tsa96 commented on src/Lumper.UI/Services/BspService.cs:
```diff
@@ -290,6 +294,30 @@ public async Task<bool> Save(IStorageFile? outFile)
                     makeBackup: MakeBackup),
                 RxApp.TaskpoolScheduler);
 
+            if (outName != FileName)
+            {
+                // update the view model to add the new keys to the UI
+                // need to do this here as otherwise the new keys will also get deleted
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateEntries(true);
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateModel();
+
+                // clean up old outdated keys
+                foreach (KeyValuePair<string, string> entry in modified)
+                {
+                    foreach (PakfileEntryViewModel pk in PakfileLumpViewModel?.Entries?.Items)
+                        if (pk.Key == entry.Key)
+                            PakfileLumpViewModel.DeleteEntry(pk);
+                }
+                // update model one more time
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateEntries(true);
+                PakfileLumpViewModel.MarkAsModified();
+                PakfileLumpViewModel.UpdateModel();
+            }
+
+
```
Shouldn't need to do this, just update keys on the viewmodel. The library code should just be updating keys on PakfileLump, which will cause the zip entries to get moved when the BSP is saved. Then you should be able to just loop through every pakfile entry viewmodel and update the key if changed. Should be all you need. Please ask on here or Discord if this stuff isn't make sense, I appreciate that the UI code can be hard to get your head around

dbampersand commented:
Oops, I totally did forget to stage that file. The UI update should be significantly simpler now.

tsa96 commented on src/Lumper.UI/Services/BspService.cs:
```diff
@@ -11,9 +11,13 @@ namespace Lumper.UI.Services;
 using System.Threading;
 using System.Threading.Tasks;
 using Avalonia.Platform.Storage;
+using Avalonia.Rendering.Composition;
```
?

tsa96 reviewed (changes requested):
Sorry it took me a while to get to this, now that my big PR is in I'll try to be more responsive with code review. This'll need rebasing on latest `origin/main`.

Note that I'll be finally running an autoformatter this codebase soon that might make rebasing hard in the future, if that's the case and you're struggling to rebase, *please* just let me know and I'd be happy to resolve everything for you.

dbampersand commented:
Thanks for the review - I've rebased onto main and I've made those changes.

tsa96 commented on src/Lumper.Lib/BSP/Lumps/BspLumps/PakFileLump.cs:
```diff
@@ -38,6 +41,91 @@ public class PakfileLump(BspFile parent) : ManagedLump<BspLumpType>(parent), IFi
 
     private static readonly Logger Logger = LogManager.GetCurrentClassLogger();
 
+    //as per https://github.com/ValveSoftware/source-sdk-2013/blob/master/mp/src/utils/vbsp/cubemap.cpp
+    //source engine only searches the following path: materials/maps/*FILENAME*/cubemapdefault(.vtf OR .hdr.vtf)
+    [GeneratedRegex(@"(materials)/(maps)/(.*)/((c-?\d+_-?\d+_-?\d+(.hdr){0,}.vtf)|(cubemapdefault(.hdr){0,}.vtf))")]
+    private static partial Regex DefaultCubemapRegex();
```
Sorry, the one I provided before was wrong, with `d` instead of `\d`. I've cleaned up this regex a bit, maybe materials/maps/ bit optional.

tsa96 commented:
Sorry I took long getting back to this, haven't had time for Lumper in a while.

I just resolved a couple things, mostly because the cubemap changes I were asking for in review didn't actually make sense, the VtfFileViewModel stuff may as well be separate, sorry about this!

Also, I noticed a major performance degradation with this, unnecessarily converting large binary files to UTF8 encoded strings to change for references to cubemap paths. Changed the UpdatePathReferences to optionally limit to a single extension, and passing in `.vmt` when called by the cubemap renaming stuff.
